### ISSUE TITLE
Firstprivate and dynamic dependency arrays as class members

### ISSUE THREAD

**Issue by Z10Frank** [bsc-pm/nanos6#9]
Hello,
in this minimal example code I have two for loops on the iterator `ibin`. Each for loop is executing `taskA` or `taskB` for the correspondent `ibin`. The execution of `taskB` on a given `ibin` is dependent on the execution of `taskA` on the same `ibin`. Thus a `firstprivate(ibin)` data sharing attribute is necessary to ensure the correctness of the procedure.

To express these dependencies, an array of dependencies `has_done_taskA[ibin]` or similar is used, to include the corresponding array element in the `depend` clause. Now, in the minimal example code this is done in 4 different ways:

1  - `bin_has_done_taskA` array is allocated statically in a function containing tasks;
2 - same as 1, but the array is defined dynamically;
3 - `bin_has_done_taskA` array is a member of a class and is allocated statically, the tasks are called within a class method;
4 - same as 4, but the array is allocated dynamically in the constructor of the class.

Methods 1,2,3 compile and work correctly, but method 4 gives this error:
```
error: the data-sharing 'firstprivate' conflicts with 'shared' required by the dependency
                        #pragma oss task firstprivate(ibin) out(bin_has_done_taskA[ibin])
                                                                                   ^~~~
```
Reading the section Data Sharing Attributes in the documentation (https://pm.bsc.es/ftp/ompss-2/doc/spec/OmpSs-2-Specification.pdf) I can't find an explanation for this different behaviour. For example, if `Dynamic storage duration objects are shared (e.g., malloc, new,...).`  as stated, shouldn't method 2 also fail?

In the code I want to taskify I use a class as in method 4, but I can easily change the procedure to method 3, hoping that allocating a dependency array each time does not worsen the performances. In any case I would like to understand better when I cannot use `firstprivate` with a dependency array.

Here is my minimal example code:
```
#include <stdio.h>
#include <iostream>

using namespace std;

void executeTasksStaticDependencyArray(int Nbin){ //compiles
	int bin_has_done_taskA[Nbin];	
        for (int ibin = 0;ibin<Nbin;ibin++){
                #pragma oss task firstprivate(ibin) out(bin_has_done_taskA[ibin])
                cout<<"task0 from ibin="<<ibin<<endl;
        }
        for (int ibin = 0;ibin<Nbin;ibin++){
                #pragma oss task firstprivate(ibin) in(bin_has_done_taskA[ibin])
                cout<<"task1 from ibin="<<ibin<<endl;
        }
        #pragma oss taskwait
}

void executeTasksDynamicDependencyArray(int Nbin){ //compiles
	int *bin_has_done_taskA;
        bin_has_done_taskA = new int[Nbin];
        for (int ibin = 0;ibin<Nbin;ibin++){
                #pragma oss task firstprivate(ibin) out(bin_has_done_taskA[ibin])
                cout<<"taskA from ibin="<<ibin<<endl;
        }
        for (int ibin = 0;ibin<Nbin;ibin++){
                #pragma oss task firstprivate(ibin) in(bin_has_done_taskA[ibin])
                cout<<"taskB from ibin="<<ibin<<endl;
        }
        #pragma oss taskwait
}


class testClass{
	public:
		testClass(int N_bins){
			Nbin=N_bins;
			bin_has_done_taskA = new int[Nbin];
		};
		~testClass(){delete bin_has_done_taskA;}

  		void classExecuteTasksStaticDependencyArray(){// compiles
			int this_bin_has_done_taskA[Nbin];
                        for (int ibin = 0;ibin<Nbin;ibin++){
                        #pragma oss task firstprivate(ibin) out(this_bin_has_done_taskA[ibin])
                        cout<<"taskA from ibin="<<ibin<<endl;
                        }
                        for (int ibin = 0;ibin<Nbin;ibin++){
                        #pragma oss task firstprivate(ibin) in(this_bin_has_done_taskA[ibin])
                        cout<<"taskB from ibin="<<ibin<<endl;
                        }

                        #pragma oss taskwait
                }

		void classExecuteTasksDynamicDependencyArray(){ // does not compile
        		
        		for (int ibin = 0;ibin<Nbin;ibin++){ // firstprivate gives an error
                	#pragma oss task firstprivate(ibin) out(bin_has_done_taskA[ibin])
               		cout<<"taskA from ibin="<<ibin<<endl;
			}
        		for (int ibin = 0;ibin<Nbin;ibin++){ // firstprivate gives an error
                	#pragma oss task firstprivate(ibin) in(bin_has_done_taskA[ibin])
                	cout<<"taskB from ibin="<<ibin<<endl;
        		}

        		#pragma oss taskwait
		}
		int Nbin;
		int *bin_has_done_taskA;
};

int main(int argc, char *argv[])
{
	int Nbin = 3;

	cout<<"Doing it with static allocation"<<endl;
        executeTasksStaticDependencyArray(Nbin);

        cout<<"Doing it with dynamic allocation"<<endl;
        executeTasksDynamicDependencyArray(Nbin);

	cout<<"Doing it with a class and static allocation"<<endl;
        testClass test_Class(Nbin);
        test_Class.classExecuteTasksStaticDependencyArray();

        cout<<"Doing it with a class and dynamic allocation"<<endl;
	test_Class.classExecuteTasksDynamicDependencyArray();

	return 0;

}
```

**Comment by mikaoP** [comment]
Hi Francesco, thank you for your report.

You're right, all these examples should compile. This is a bug we fixed in our private branch. At this point in time we published the most recent changes, so I would recommend you to update your installation to include the fix. The specific commit that faces this issue is https://github.com/bsc-pm/llvm/commit/ecc7282a0f7f8494366e42dbc710ffda388ccec9

Regarding your data-sharing question, all those cases would need to refer to a shared object. The difference is that in the examples 2 and 4 the variables are pointers. As you noticed in the spec, these dynamic objects have to be shared, and what is done in practice is to firstprivatize the pointer, in contrast to variables that are not pointers which are shared.
